Make traffic-agents configuration follow suit when things change

thallgren opened the pull request:
## Description

A service that is removed or added will often affect the set of possible
routes that can be intercepted. This commit adds a watcher for services
and logic to discover what existing entries that may be affected by
the service addition/modification/deletion.

In addition, the mutating webhook now also listens for pod deletions, so
that the corresponding configmap entry can be removed when the owner
workload cease to exist.

## Checklist

 - [x] I made sure to update `./CHANGELOG.md`.
 - [x] My change is adequately tested.

Commit:
Make intercept state know about all possible intercepts

An intercept state must know about all possible intercepts (i.e. the
intercepts using the same port) in order to be matchable against an
intercept specification.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Update changelog.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Make traffic-agents configuration follow suit when things change

A service that is removed or added will often affect the set of possible
routes that can be intercepted. This commit adds a watcher for services
and logic to discover what existing entries that may be affected by
the service addition/modification/deletion.

In addition, the mutating webhook now also listens for pod deletions, so
that the corresponding configmap entry can be removed when the owner
workload cease to exist.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

josecv commented on cmd/traffic/cmd/manager/internal/mutator/watcher.go:
```diff
@@ -205,6 +245,15 @@ func (c *configWatcher) Store(ctx context.Context, ac *agentconfig.Sidecar, upda
+		if currentYml, ok := cm.Data[ac.AgentName]; ok {
+			var currAc agentconfig.Sidecar
+			if err = decode(currentYml, &currAc); err == nil && currAc.Manual {
+				dlog.Warn(ctx, "avoided an attempt to overwrite manually added config entry for %s.%s", ac.AgentName, ns)
```
I think this should be warnf since you have a formatting directive at the end there

Commit:
Fix dlog.Warn -> dlog.Warnf

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Fix problem with watcher not terminating correctly when cancelled

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Fix problem uninstall --agent nnn or --all-agents

A cluster-wide watcher would be started when no namespace was given.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
chart

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Fix problem with config being recreated by webhook delete event.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Ensure proper cleanup in integration_tests

The other-NNN namespace and the Helm install in the custom-NNN namespace
wasn't cleaned up properly.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Make daemon session terminate properly when traffic-manager dies

A Helm uninstall of the traffic-manager left the root daemon in a bad
state, constantly trying to reconnect to obtain cluster info. This
commit ensures that the session dies.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>
